### PR TITLE
V2: improve CLI for multitask

### PR DESCRIPTION
**Pull request by hwpang**
## Description
- [x] Fix some small bugs related to multitask training and prediction.
- [x] Add a test dataset for single molecule multitask regression
- [x] Add a CLI test for training and predicting multitask
- [x] Improve saving predictions for
    - [x] multi-task
    - multi-task MVE model -> part of v2.1, won't support for now for v2.0.
    - [x] multi-task multiclass -> looking for an example dataset to test this out. -> addressed in https://github.com/chemprop/chemprop/pull/672.

While implementing the cli test for multi-task MVE model, I have encountered bugs related to MVE outside the scope of this PR. I will make another PR to fix it. Edit: single-task MVE is fixed in https://github.com/chemprop/chemprop/pull/664.
 
## Question
- Do we have an example multi-class dataset?

## Checklist
- [x] linted with flake8?
- [x] (if appropriate) unit tests added?

### CODE REVIEW

File: chemprop/models/model.py [hunk 1]
```diff
@@ -173,6 +173,10 @@ def _evaluate_batch(self, batch) -> list[Tensor]:
         targets = targets.nan_to_num(nan=0.0)
         preds = self(bmg, V_d, X_f)
 
+        if isinstance(self.predictor, MveFFN):
+            mean, _ = torch.chunk(preds, self.predictor.n_targets, 1)
+            preds = mean
```
Comment (KnathanM) [inline chemprop/models/model.py:178]: If I understand correctly, you add this type check here because an `MveFFN` predicts two targets, `mean` and `var` and when computing a metric we only need the mean. I see `EvidentialFFN`s also predict multiple things (mean, v, alpha, beta). I assume we would also need to include a type check for `EvidentialFFN`s. 

We could do that, but it may be better to require `metric` to take the whole raw `preds` and internally split it into the components. That is what the loss functions do (see `MVELoss` and `EvidentialLoss`). I notice many of the metric classes in `metrics.py` just inherit from their corresponding loss and then change nothing, so that they are really just an alias for the loss class. I think we should do that for MVELoss and EvidentialLoss and include them as the `_default_metric` for MveFFN and EvidentialFFN. Then we don't need the type checking here as `mertic` should take care of spliting the predictions.

File: chemprop/nn/predictors.py [hunk 2]
```diff
@@ -120,8 +120,17 @@ def __init__(
...
+            loc = torch.ones(1, self.n_tasks) * loc
+        else:
+            loc = torch.tensor(loc).view(1, -1)
+        self.register_buffer("loc", loc)
+
+        if not isinstance(scale, Tensor):
+            scale = torch.ones(1, self.n_tasks) * scale
+        else:
+            scale = torch.tensor(scale).view(1, -1)
+        self.register_buffer("scale", scale)
```
Comment (KnathanM) [inline chemprop/nn/predictors.py:133]: I see that the CLI doesn't use give the scaler to `build_model()` which means that when making a `RegressionFFN` the `loc` and `scale` use default values of 0 and 1. Later in the CLI we set these manually (`model.predictor.register_buffer("loc", torch.tensor(scaler.mean_).view(1, -1))`). But if the user was building their own `RegressionFFN`, they would use this code. It looks like this code is written to handle the use case of passing a pair of floats loc/scale to transform multiple tasks. I think if the user has multiple tasks, they will have a different loc/scale for each target and we can expect it to be an array. (So the original code would be fine?)

This also reminds me that we wanted to spend time thinking about how we load and save scalers in general.

File: chemprop/models/model.py (hunk 1 above)
Comment (hwpang) [reply chemprop/models/model.py:178]: That make sense! I split this part of changes relevant to `MveFFN` into https://github.com/chemprop/chemprop/pull/664. I will work on them later as they are more in the scope of v2.1.

File: chemprop/nn/predictors.py (hunk 2 above)
Comment (hwpang) [reply chemprop/nn/predictors.py]: This part of code is to handle error I ran into when loading a trained multitask model. The following lines initialize the `loc` and `scale` with default values of 0 and 1 and turn them into a 1 by 1 tensor respectively, but the trained model contains a different shape of tensor (1 by number of tasks), so it causes pytorch to throw error when writing in the weights.

```
self.register_buffer("loc", torch.tensor(loc).view(1, -1))
self.register_buffer("scale", torch.tensor(scale).view(1, -1))
```

I see what you're saying. I will change the type check so that it only handles the case when `loc` and `scale` are floats.

**Comment by KnathanM** [comment]
I now see that the test are failing. 

In `tests/cli/test_cli_regression_mol.py` you have `test_train_quick` defined twice. The second one uses `regression-mve` which we are moving to v2.1 so I think you can just delete the second `test_train_quick` to fix it.

The second error seems a bit more difficult. In `tests/cli/test_cli_regression_mol_multitask.py` you try to load a multitask model but the model seems to have loc and scale initialized as 1x1 torch tensors, so overwriting them with the 1x12 tensors in the model file throws an error. Maybe the model's `ntasks` attribute needs to be set sooner so the model will be initialized with a 1x12 loc and scale?

**Comment by hwpang** [comment]
I took a look into this. The error is the same that I saw previously: when `loc` is initialized to be a 1 by 1 tensor but the trained model is 1 by n_tasks, it has this error.

I found that the `isinstance(loc, float)` type check is not working as expected. For some reason, the default `loc` is recognized as an integer. Changing `loc` to 0.0 doesn't work. I can change the type check to check for int instead, but it can be confusing.

File: chemprop/models/model.py [hunk 3]
```diff
@@ -10,7 +10,7 @@
 
 from chemprop.data import TrainingBatch, BatchMolGraph
 from chemprop.nn.metrics import Metric
-from chemprop.nn import MessagePassing, Aggregation, Predictor, LossFunction
+from chemprop.nn import MessagePassing, Aggregation, Predictor, LossFunction, MveFFN
```
Comment (KnathanM) [inline chemprop/models/model.py:13]:
```suggestion
from chemprop.nn import MessagePassing, Aggregation, Predictor, LossFunction
```
This isn't used now that you moved those changes to another PR. Last fix and then I will merge this PR in.

**Review by KnathanM — APPROVED** [review]
